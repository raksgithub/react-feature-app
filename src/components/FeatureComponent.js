import React from 'react';
import Proptypes from 'prop-types';

function makeStyleOfContainer(title, bought, by) {
  if (bought && title !== by) {
    return {
      backgroundColor: '#d1d0cd'
    }
  }
  switch (title) {
    case 'starter':
      return { backgroundColor: '#868a83' };
    case 'regular':
      return { backgroundColor: '#1a94d6' };
    case 'professional':
      return { backgroundColor: '#6f75a6' };
    default:
      return { backgroundColor: '#fa3628' };
  }
}

function makeStyleOfButton(title, bought, by) {
  if (bought && title !== by) {
    return {
      backgroundColor: '#e3e3dc',
      cursor: 'default',
      color: 'white'
    }
  }
}

export default function FeatureComponent(props) {
  const { features, bought, by, handleButtonAction } = props;

  return (
    <div className="feature-container">
      {features.map(feature => (
        <div
          className="feature-item"
          key={feature.id}
          style={makeStyleOfContainer(feature.title, bought, by)}
        >
          <div className="feature-item-title">
            {feature.title.toUpperCase()}
          </div>
          <div className="feature-item-price">
            {bought && by === feature.title ? 'BOUGHT!' : feature.price}
          </div>
          <div className="feature-item-footer">
            <p>{feature.description}</p>
            <button
              onClick={() => handleButtonAction(feature.title)}
              disabled={bought && by !== feature.title}
              style={makeStyleOfButton(feature.title, bought, by)}
              className="button"
            >
              {bought && by === feature.title ? 'CANCEL' : 'BUY'}
            </button>
          </div>
        </div>
      ))}
    </div>
  );
}

FeatureComponent.propTypes = {
  features: Proptypes.arrayOf(Proptypes.object).isRequired,
  bought: Proptypes.bool.isRequired,
  by: Proptypes.string,
  handleButtonAction: Proptypes.func.isRequired
};