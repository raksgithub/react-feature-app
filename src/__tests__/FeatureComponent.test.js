import React from 'react';
import ReactDOM from 'react-dom';
import { create } from 'react-test-renderer';
import FeatureComponent from '../components/FeatureComponent/';

let container = null;

// Setup Logic
beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

describe('Test FeatureComponent with Snapshot testing', () => {
    it('Tests One Component features when item is not bought by anyone', () => {
        const features = [{
            id: 1,
            title: 'starter',
            price: '$1',
            description: 'Starter features for your business to grow.'
        }, {
            id: 2,
            title: 'regular',
            price: '$25',
            description: 'Regular features for your business to grow.'
        }];
        const bought = false;
        const by = 'null';
        const handleButtonAction = function() {};
        const component = create(
            <FeatureComponent 
                features={features} 
                bought={bought} 
                by={by} 
                handleButtonAction={handleButtonAction} 
            />
        );
        // const instance = component.getInstance();
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Tests One Component features when item is bought by starter', () => {
        const features = [{
            id: 1,
            title: 'starter',
            price: '$1',
            description: 'Starter features for your business to grow.'
        }, {
            id: 2,
            title: 'regular',
            price: '$25',
            description: 'Regular features for your business to grow.'
        }];
        const bought = true;
        const by = 'starter';
        const handleButtonAction = function() {};
        const component = create(
            <FeatureComponent 
                features={features} 
                bought={bought} 
                by={by} 
                handleButtonAction={handleButtonAction} 
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Tests Two Component features when item is not bought by anyone', () => {
        const features = [{
            id: 1,
            title: 'professional',
            price: '$75',
            description: 'Professional features for your business to grow.'
        }, {
            id: 2,
            title: 'ultimate',
            price: '$115',
            description: 'The ultimate set of features for your business to grow.'
        }];
        const bought = false;
        const by = 'null';
        const handleButtonAction = function() {};
        const component = create(
            <FeatureComponent 
                features={features} 
                bought={bought} 
                by={by} 
                handleButtonAction={handleButtonAction} 
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Tests One Component features when item is bought by ultimate', () => {
        const features = [{
            id: 1,
            title: 'professional',
            price: '$75',
            description: 'Professional features for your business to grow.'
        }, {
            id: 2,
            title: 'ultimate',
            price: '$115',
            description: 'The ultimate set of features for your business to grow.'
        }];
        const bought = true;
        const by = 'ultimate';
        const handleButtonAction = function() {};
        const component = create(
            <FeatureComponent 
                features={features} 
                bought={bought} 
                by={by} 
                handleButtonAction={handleButtonAction} 
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});

// Teardown Logic
afterEach(() => {
    ReactDOM.unmountComponentAtNode(container);
    container.remove();
    container = null;
});
