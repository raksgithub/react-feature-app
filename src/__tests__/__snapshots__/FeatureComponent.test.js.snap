// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FeatureComponent with Snapshot testing Tests One Component features when item is bought by starter 1`] = `
<div
  className="feature-container"
>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#868a83",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      STARTER
    </div>
    <div
      className="feature-item-price"
    >
      BOUGHT!
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        Starter features for your business to grow.
      </p>
      <button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        CANCEL
      </button>
    </div>
  </div>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#d1d0cd",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      REGULAR
    </div>
    <div
      className="feature-item-price"
    >
      $25
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        Regular features for your business to grow.
      </p>
      <button
        className="button"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#e3e3dc",
            "color": "white",
            "cursor": "default",
          }
        }
      >
        BUY
      </button>
    </div>
  </div>
</div>
`;

exports[`Test FeatureComponent with Snapshot testing Tests One Component features when item is bought by ultimate 1`] = `
<div
  className="feature-container"
>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#d1d0cd",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      PROFESSIONAL
    </div>
    <div
      className="feature-item-price"
    >
      $75
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        Professional features for your business to grow.
      </p>
      <button
        className="button"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#e3e3dc",
            "color": "white",
            "cursor": "default",
          }
        }
      >
        BUY
      </button>
    </div>
  </div>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#fa3628",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      ULTIMATE
    </div>
    <div
      className="feature-item-price"
    >
      BOUGHT!
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        The ultimate set of features for your business to grow.
      </p>
      <button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        CANCEL
      </button>
    </div>
  </div>
</div>
`;

exports[`Test FeatureComponent with Snapshot testing Tests One Component features when item is not bought by anyone 1`] = `
<div
  className="feature-container"
>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#868a83",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      STARTER
    </div>
    <div
      className="feature-item-price"
    >
      $1
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        Starter features for your business to grow.
      </p>
      <button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        BUY
      </button>
    </div>
  </div>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#1a94d6",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      REGULAR
    </div>
    <div
      className="feature-item-price"
    >
      $25
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        Regular features for your business to grow.
      </p>
      <button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        BUY
      </button>
    </div>
  </div>
</div>
`;

exports[`Test FeatureComponent with Snapshot testing Tests Two Component features when item is not bought by anyone 1`] = `
<div
  className="feature-container"
>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#6f75a6",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      PROFESSIONAL
    </div>
    <div
      className="feature-item-price"
    >
      $75
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        Professional features for your business to grow.
      </p>
      <button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        BUY
      </button>
    </div>
  </div>
  <div
    className="feature-item"
    style={
      Object {
        "backgroundColor": "#fa3628",
      }
    }
  >
    <div
      className="feature-item-title"
    >
      ULTIMATE
    </div>
    <div
      className="feature-item-price"
    >
      $115
    </div>
    <div
      className="feature-item-footer"
    >
      <p>
        The ultimate set of features for your business to grow.
      </p>
      <button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        BUY
      </button>
    </div>
  </div>
</div>
`;
